test/upgrade/main.go

```diff
@@ -104,50 +136,164 @@ func findK8sVersion() string {
+// kubectl create namespace agones-system
+// kubectl apply --server-side -f https://raw.githubusercontent.com/googleforgames/agones/release-1.43.0/install/yaml/install.yaml
+func installYaml() error {
+	installArgs := []string{"apply", "--server-side", "-f", "https://raw.githubusercontent.com/googleforgames/agones/release-1.43.0/install/yaml/install.yaml"}
+
+	_, err := runExecCommand(KUBECTL_CMD, installArgs...)
+	return err
+}
+
+// Adds public Helm Agones releases to the cluster
+func addAgonesRepo() {
+	installArgs := [][]string{{"repo", "add", "agones", "https://agones.dev/chart/stable"},
+		{"repo", "update"}}
+
+	for _, args := range installArgs {
+		_, err := runExecCommand(HELM_CMD, args...)
+		if err != nil {
+			log.Fatalf("Could not add Agones helm repo: %s", err)
 		}
 	}
+}
+
+func installAgonesRelease(version, registry, featureGates, imagePullPolicy, sidecarPullPolicy,
+	logLevel, chart string) error {
+
+	// TODO: Include feature gates. (Current issue with Helm and string formatting of the feature gates.)
+	// 		"--set agones.featureGates=%s "+
+
+	helmString := fmt.Sprintf(
+		"upgrade --install --atomic --wait --timeout=10m --namespace=agones-system --create-namespace --version %s "+
+			"--set agones.image.tag=%s "+
+			"--set agones.image.registry=%s "+
+			"--set agones.image.allocator.pullPolicy=%s "+
+			"--set agones.image.controller.pullPolicy=%s "+
+			"--set agones.image.extensions.pullPolicy=%s "+
+			"--set agones.image.ping.pullPolicy=%s "+
+			"--set agones.image.sdk.alwaysPull=%s "+
+			"--set agones.controller.logLevel=%s "+
+			"agones %s",
+		version, version, registry, imagePullPolicy, imagePullPolicy, imagePullPolicy, imagePullPolicy,
+		sidecarPullPolicy, logLevel, chart,
+	)
+
+	helmArgs := strings.Split(helmString, " ")
+
+	_, err := runExecCommand(HELM_CMD, helmArgs...)
+	if err != nil {
+		return err
+	}
+
+	return err
+}
+
+func runConfigWalker(ctx context.Context, validConfigs []*configTest) {
+	for _, config := range validConfigs {
+		registry := AGONES_REGISTRY
+		chart := HELM_CHART
+		if config.agonesVersion == DEV {
+			// TODO: Update to templated value for registry and chart for Dev build
+			continue
+		}
 
-	for i, featureGate := range fg.BetaGates {
-		if i < (len(fg.BetaGates) - 1) {
-			fmt.Fprintf(&featureWithGate, "%s=true&", featureGate)
-		} else {
-			fmt.Fprintf(&featureWithGate, "%s=true", featureGate)
+		err := installAgonesRelease(config.agonesVersion, registry, config.featureGates, IMAGE_PULL_POLICY,
+			SIDECAR_PULL_POLICY, LOG_LEVEL, chart)
+		if err != nil {
+			log.Printf("installAgonesRelease err: %s", err)
 		}
+
+		// Wait for the helm release to install. Waits the same amount of time as the Helm timeout.
+		var helmStatus string
+		err = wait.PollUntilContextTimeout(ctx, 10*time.Second, 10*time.Minute, true, func(ctx context.Context) (done bool, err error) {
+			helmStatus = checkHelmStatus(config.agonesVersion)
+			if helmStatus == "deployed" {
```

Comment:
I have some trouble to understand the helm doc so I'll just ask. Does a "deployed" helm release necessarily mean the underlying k8s pods are running?

Reply:
In this case it does, because we run helm with the argument `wait`, which waits until all the underlying pods are running before marking the release as `deployed`.